libraries/Microsoft.Bot.Builder.Dialogs.Rules/Steps/HttpRequest.cs

```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.Dialogs.Rules.Steps
+{
+    public class HttpRequest : DialogCommand
+    {
+
+        public enum Method {
+            GET,
+            POST
+        }
+
+        public HttpRequest()
+            : base()
+        {
+        }
+
+        protected override string OnComputeId()
+        {
+            return $"HttpRequest[{method}]";
+        }
+
+        public Method method;
+
+        public string url;
+
+        public string responseProperty;
+
+        public Dictionary<string, string> header;
+
+        public JObject body;
+
+        public HttpRequest(Method method, string url, string responseProperty, Dictionary<string, string> header = null, JObject body = null)
+        {
+
+            this.method = method;
+            this.url = url;
+            this.responseProperty = responseProperty;
+            if (header != null)
+            {
+                this.header = header;
+            }
+            
+            if (body != null)
+            {
+                this.body = body;
+            }
+        }
+
+        protected override async Task<DialogTurnResult> OnRunCommandAsync(DialogContext dc, object options = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            this.url = await new TextTemplate(this.url).BindToData(dc.Context, dc.State, (property, data) => dc.State.GetValue<object>(data, property)).ConfigureAwait(false);
+
+            // Bind each string token to the data in state
+            if (body != null)
+            {
+                foreach (var unit in this.body)
+                {
+                    if (unit.Value.Type == JTokenType.String)
+                    {
+                        unit.Value.Replace(await new TextTemplate(unit.Value.ToString()).BindToData(dc.Context, dc.State, (property, data) => dc.State.GetValue<object>(data, property)));
+                    }
+                }
+            }
+
+            var client = new HttpClient();
```

Comment:
We need to be more careful when managing HttpClients as they implement IDisposable and are bound to I/O resources such as connections. Maybe have a global http client and pass it in or a static http client in this class or at least have an instance httpClient.

Reply:
fixed